src/librustc_mir/build/mod.rs

```diff
@@ -883,21 +881,13 @@ impl<'a, 'gcx, 'tcx> Builder<'a, 'gcx, 'tcx> {
             // debuginfo and so that error reporting knows that this is a user
             // variable. For any other pattern the pattern introduces new
             // variables which will be named instead.
-            let mut name = None;
-            if let Some(pat) = pattern {
-                match pat.node {
-                    hir::PatKind::Binding(hir::BindingAnnotation::Unannotated, _, ident, _)
-                    | hir::PatKind::Binding(hir::BindingAnnotation::Mutable, _, ident, _) => {
-                        name = Some(ident.name);
-                    }
-                    _ => (),
-                }
-            }
-
-            let source_info = SourceInfo {
-                scope: OUTERMOST_SOURCE_SCOPE,
-                span: pattern.map_or(self.fn_span, |pat| pat.span)
+            let (name, span) = if let Some(pat) = pattern {
+                (pat.simple_ident().map(|ident| ident.name), pat.span)
```

Comment:
@pnkfelix @nikomatsakis @matthewjasper Is this correct, as far as you know?
(if nobody replies I expect we can just merge without this change, this only affects debuginfo, and not in a significant way.

Reply:
I think that the corresponding change needs to be made on line 925 below. Or this should be combined with that check.

Reply:
@matthewjasper could you clarify on what corresponding change needs to be made? 

I'm probably misunderstanding what you're referring to, but as far as I can tell the `BindingMode::ByValue` checked on line 925 is equivalent to `BindingAnnotation::Unannotated` and `BindingAnnotation::Mutable` that `simple_ident` checks (and that the code this PR changes checked for)?

Reply:
You also need to check that there's no subpattern.

Reply:
Ah, hadn't spotted that, fixed, thanks!